File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
'''
(와 )는 짝궁이어야함
-> 닫는 괄호가 왔을 때, 여는 괄호가 없으면 오류, 다 지났는데 괄호가 안닫혀도 오류

( 가 오면 스택에 push )가 오면 스택에서 pop
)가 왔는데 스택이 empty이면 오류
다 순회했는데 스택이 empty가 아니면 오류
'''
def solution(s):
    stack = []
    #print(stack)
    for v in s:
        if v == ')':
            if stack:
                stack.pop()
            else:
                return False
        else:
            stack.append(0)
    if stack:
        return False
    return True
                
            
            

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;
class Solution {
    boolean solution(String s) {
        Stack<Character> stack = new Stack(); 
        /*
        여는 괄호가 나오면 stack에 Push
        닫는 괄호가 나오면 pop
        닫는 괄호가 나왔는데 stack이 empty이면 false
        마지막에 stack이 남아 있는채로 끝나도 false
        */
        for(int i=0; i<s.length(); i++){
            if(s.charAt(i) == ')'){
                if(stack.isEmpty()){
                    return false;
                }
                else{
                    stack.pop();
                }
            }
            else{
                stack.push('(');
            }
        }
        // 닫히지 않는 괄호가 있는 경우
        if(!stack.isEmpty()){
            return false;
        }
        return true;
    }
}